test/TensorFlowRuntime/dataset_api.swift

```diff
@@ -0,0 +1,52 @@
+// RUN: %target-run-simple-swift
+// REQUIRES: executable_test
+// REQUIRES: swift_test_mode_optimize
+//
+// Dataset API tests.
+
+import TensorFlow
+
+#if TPU
+import TensorFlowUnittestTPU
+#else
+import TensorFlowUnittest
+#endif
+import StdlibUnittest
+
+var DatasetAPITests = TestSuite("DatasetAPI")
+
+DatasetAPITests.testAllBackends("SingleValueManualIterator") {
+  // [[1], [2], [3], [4], [5]]
+  let scalars = Tensor<Float>(rangeFrom: 0, to: 5, stride: 1).reshaped(to: [5, 1])
+  let dataset = SingleValueDataset(elements: scalars, elementShape: [1])
+  var iterator = dataset.makeIterator()
+  var i: Int32 = 0
+  while let item = iterator.next() {
+    expectEqual(scalars[i].array, item.array)
+    i += 1
+  }
+}
+
+DatasetAPITests.testAllBackends("SingleValueDatasetIteration") {
+  // [[1], [2], [3], [4], [5]]
+  let scalars = Tensor<Float>(rangeFrom: 0, to: 5, stride: 1).reshaped(to: [5, 1])
+  let dataset = SingleValueDataset(elements: scalars, elementShape: [1])
+  var i: Int32 = 0
+  for item in dataset {
```

Comment:
i'm not sure if the support of iterating over dataset could be a useful short-hand, or could lead to confusion.

should we stick to the TF dataset API design, where user must first create an iterator out of a dataset?

Reply:
You seem to suggest that TensorFlow does not support iterating over, but that’s not the case. TensorFlow’s data API *does* support being iterated over. Python’s tf.data.Dataset conforms to Python’s __iter__ protocol and *can* be iterated over. 

Plus, the entire dataset is consistent with standard Swift protocols both semantics-wise and API-wise. I don’t think I can see any arguments for confusion from your speculation.

Reply:
From https://www.tensorflow.org/guide/datasets, we have:

> Once you have built a Dataset to represent your input data, the next step is to create an Iterator to access elements from that dataset.

If we don't make dataset conform to sequence, the benefits are:

1. allow us to customize the makeIterator() API, and support things like make_one_shot_iterator and make_initializable_iterator if needed (see https://www.tensorflow.org/guide/datasets).

2. remove the need to define what it means to iterate over a dataset. is it a short hand for creating a one shot iterator? can i iterate over a dataset multiple times with a sequential list of for-in loops? can i create multiple threads that concurrently iterate over the same dataset object?

It seems simpler conceptually, if we define a dataset to be a stateless object -- a simple descriptor of how we read and transform the data  (e.g. from file, and apply shuffling, etc). To do the actual reading, create an iterator, which is stateful and not thread-safe. The use cases can then be:
-  if you want to iterate over the same dataset in multiple concurrent threads, where each thread reads its entire copy of data, create one iterator for each such thread. no synchronization is needed.
- if you want multiple threads to collaborate in reading from the same iterator, use a mutex to protect that iterator get_next calls. alternatively, have a single thread drive the get_next calls, and then dispatch/multiplex the output to multiple consumers.

Reply:
> 1. allow us to customize the makeIterator() API, and support things like make_one_shot_iterator and make_initializable_iterator if needed (see https://www.tensorflow.org/guide/datasets).

Yes, different iterators will be supported and I have considered this. This is not a benefit of not conforming to `Sequence`; this is a benefit of adding more APIs.

```swift
extension SingleValueDataset {
  func makeOneShotIterator() -> OneShotDatasetIterator<Scalar>
  func makeInitializableIterator() -> InitializableDatasetIterator<Scalar>
}
```

But a sequence like `tf.Dataset` has well-defined semantics for [what it means to iterator over elements a dataset](https://github.com/tensorflow/tensorflow/blob/25c197e02393bd44f50079945409009dd4d434f8/tensorflow/python/data/ops/dataset_ops.py#L113). That is why iterating over a dataset is very naturally expressed like iterating over a sequence in TensorFlow Eager. The user certainly has the ability to choose what iterators they use.

> 2. remove the need to define what it means to iterate over a dataset. is it a short hand for creating a one shot iterator?

It creates a default iterator for traversing a sequence, like Python's `__iter__` which internally calls `MakeIterator`. 

A one-shot iterator is the most flexible API because it allows the user to specify a generator. This API is going to exist and is orthogonal to whether `SingleValueDataset` conforms to `Sequence`. It can either be modeled as a separate type like Swift's [`AnyIterator`](https://developer.apple.com/documentation/swift/anysequence/2854368-init), or can just be created using a different `init` that uses `OneShotIterator` op: 

```swift
extension SingleValueDatasetIterator {
  init(datasetFactory: @convention(tensorflow) () -> SingleValueDataset<Scalar>)
}
```

The underlying mechanism is analogous to `datasetFactory().makeIterator()`, except that this makes the dataset closed over by the closure and thereby anonymous.

> can i iterate over a dataset multiple times with a sequential list of for-in loops? 

Yes, this is the fundamental semantics of TensorFlow's Dataset object. It is pure and can be iterated over multiple times.

> can i create multiple threads that concurrently iterate over the same dataset object?

By value semantics of Dataset, certainly yes. If you ask in a context of distributed programming, this is a also general question for TensorFlow's `Dataset` API. In both worlds, `Dataset` is stateless and `Iterator` is stateful.

> It seems simpler conceptually, if we define a dataset to be a stateless object -- a simple descriptor of how we read and transform the data (e.g. from file, and apply shuffling, etc). To do the actual reading, create an iterator, which is stateful and not thread-safe.

This *is* the current design. `Dataset` already is a stateless object by value semantics and it can be iterated over multiple times. The semantics is exactly the same as TensorFlow's `Dataset`.

You seem to believe that a `for`-`in` loop mutates the dataset, is that right? That's not the case because in Python or Swift because, a `for`-`in` loop creating an stateful iterator internally to iterate over. It is not mutating the dataset.

> * if you want to iterate over the same dataset in multiple concurrent threads, where each thread reads its entire copy of data, create one iterator for each such thread. no synchronization is needed.
> * if you want multiple threads to collaborate in reading from the same iterator, use a mutex to protect that iterator get_next calls. alternatively, have a single thread drive the get_next calls, and then dispatch/multiplex the output to multiple consumers.

Correct. With value semantics of `Dataset`, this is the exact behavior that this API offers. This is completely orthogonal to whether `Dataset` conforms to `Sequence`. Note that Swift's `Sequence` does not mean that the conforming object is stateful. It means that 1) an iterator can be created for that object and 2) there are functional operators the can be applied to the object to transform the sequence. This is exactly consistent with TensorFlow's `Dataset` and `Iterator` design. So I believe there's a misunderstanding here.

Reply:
Ok, thanks for the various context. It seems we are agreeing on the same basic design, and the iteration support over a dataset is just syntax sugar that internally creates an iterator obj for each for-in loop. This is fine.

A few suggestions:
1. It'd be nice to not have the cardinality info leaked into on both dataset and iterator structs (returning a `OneShotDatasetIterator<Scalar>` is too much detail)
2. Can we model cardinality via an array of meta types instead of template args?

The rough design would look like:
```swift
protocol Dataset {
  // DatasetIterator is a protocol, so that user need not deal with concrete types of iterators.
  func makeOneShotIterator() -> DatasetIterator
  func makeInitializableIterator() -> DatasetIterator
  ...
  // if we want to conform to Sequence protocol, let makeIterator() map to one of the above make methods?
}

struct FileDataset : Dataset {
  // an init method over a file path
  public init(types: [AcceleratorByTensorFlow], shapes: [TensorShape], filePath: String) {
    ...
  }
  // implement the protocol methods from Dataset
  // e.g. makeOneShotIterator() creates an iterator based on the types and shapes known to the dataset.
}

struct TensorSliceDataset : Dataset {
  // an init method over a tensor
  // implement the protocol methods from Dataset
}
...
```

Reply:
I don't see a need for variadic generics here.  Why can't you use structs as suggested above, or use tuple types, e.g.:

```
typealias SingleValueDataset<T> = Dataset<T>
typealias DoubleValueDataset<T, U> = Dataset<(T, U)>
typealias TripleValueDataset<T, U, V> = Dataset<(T, U, V)>
typealias QuadrupleValueDataset<T, U, V, W> = Dataset<(T, U, V, W)>
```

The only downside to using tuples like I showed above is that tuples can't conform to protocols, so you can't constrain the element type of DataSet.  That said, structs should work fine and solve that problem.

Reply:
> The only downside to using tuples like I showed above is that tuples can't conform to protocols.

That's what I meant by "variadic tuples" and I should've made that clearer. I believe the consensus on Swift evolution was that variadic tuples and tuples being an alias to `Tuple<T...>` are a prerequisite for making tuples able to conform to protocols.

I showed some problems with your struct approach to handle multiple results in a separate comment. I would appreciate ideas on the tactics to resolve this. Again, I'm also not a fan of having multiple types representing datasets with different arity, but the technical constraint with `#tfop` is a real bottleneck.